### PR TITLE
Bringup cache updates to ttir level

### PR DESCRIPTION
**Pull request by LPanosTT**
Add `ttnn.update_cache`, `ttir.update_cache`, `ttir.fill_cache`, and `ttnn.fill_cache`

- Frameworks store cache position in a tensor, ttnn needs a `uint32_t` update index (not a tensor). So In runtime I currently read this value from the tensor buffer and call the op. Made a runtime workaround and filed an issue: https://github.com/tenstorrent/tt-mlir/issues/1510
- I noticed that `ttir.constant` ops where the value is 0 is lowered to `ttnn.empty`. I've made a test where the update index tensor holds a single zero. In the runtime however, the tensor returned by the `ttnn.empty` did not have 0 in it, it had a random integer. So I've changed the `ttir.constant` lowering to just use `ttnn.full` no matter the value.
    - Needed to change the file checks for some tests with `ttir.constant`

### CODE REVIEW

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td [hunk 1]
```diff
@@ -693,6 +693,25 @@ def TTIR_ConcatOp : TTIR_DPSOp<"concat"> {
     let hasVerifier = 1;
 }
 
+def TTNN_UpdateCacheForTokenOp : TTIR_DPSOp<"update_cache_for_token"> {
```
Comment (nsmithtt) [inline include/ttmlir/Dialect/TTIR/IR/TTIROps.td:696]: Typo `TTNN` -> `TTIR`

File: runtime/lib/ttnn/operations/kv_cache/fill_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "fill_cache.h"
+#include "tt/runtime/detail/logger.h"
+#include <ttnn/operations/core/core.hpp>
```
Comment (jnie-TT) [inline runtime/lib/ttnn/operations/kv_cache/fill_cache.cpp:7]: These should be in quotations, also do we need these includes?

File: runtime/lib/ttnn/operations/kv_cache/update_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "update_cache.h"
+#include "tt/runtime/detail/logger.h"
+#include <ttnn/operations/core/core.hpp>
```
Comment (jnie-TT) [inline runtime/lib/ttnn/operations/kv_cache/update_cache.cpp:7]: Same here

File: runtime/lib/ttnn/operations/kv_cache/update_cache.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  const ::ttnn::Tensor &input = tensorPool.at(op->input()->global_id());
+  const ::ttnn::Tensor &updateIndex =
+      tensorPool.at(op->update_index()->global_id());
+
+  const ::ttnn::Tensor index_on_host = ::ttnn::from_device(updateIndex);
+  const auto storage = index_on_host.get_storage();
+  const auto ownedStorage = std::get<tt_metal::OwnedStorage>(storage);
+  const auto buffer = ownedStorage.get_buffer();
+  const auto buf = std::get<tt_metal::owned_buffer::Buffer<uint32_t>>(buffer);
+  uint32_t up_idx = *buf.begin();
```
Comment (jnie-TT) [inline runtime/lib/ttnn/operations/kv_cache/update_cache.cpp:25]: We should probably find a better way to do this, or maybe mark it as a workaround

File: runtime/lib/ttnn/operations/kv_cache/update_cache.cpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+  ProgramTensorPool &tensorPool = context.getTensorPool();
+
+  const ::ttnn::Tensor &cache = tensorPool.at(op->cache()->global_id());
+  const ::ttnn::Tensor &input = tensorPool.at(op->input()->global_id());
+  const ::ttnn::Tensor &updateIndex =
+      tensorPool.at(op->update_index()->global_id());
+
+  const ::ttnn::Tensor index_on_host = ::ttnn::from_device(updateIndex);
+  const auto storage = index_on_host.get_storage();
```
Comment (jnie-TT) [inline runtime/lib/ttnn/operations/kv_cache/update_cache.cpp:21]: nit: camelCase

**Comment by mmanzoorTT** [comment]
FYI @mrakitaTT. This PR converts `ttir.constant<0>` to `ttnn.full<0>`

**Review by jnie-TT — APPROVED** [review]

File: include/ttmlir/Dialect/TT/IR/TTOpsTypes.h [hunk 6]
```diff
@@ -110,6 +110,37 @@ inline DataType elementTypeToDataType(Type elementType) {
...
+  case DataType::UInt32:
+    return "UInt32";
+  case DataType::UInt16:
+    return "UInt16";
+  case DataType::UInt8:
+    return "UInt8";
+  }
+  llvm_unreachable("unhandled data type");
+}
+
```
Comment (nsmithtt) [inline include/ttmlir/Dialect/TT/IR/TTOpsTypes.h:143]: Tablegen autogens this, `::llvm::StringRef DataTypeEnumToString(DataType);`

**Review by nsmithtt — APPROVED** [review]
Hey Lewis this looks great!  I presume the `deallocate` of the kv cache is still an issue.  How are we working around that?

File: runtime/lib/ttnn/operations/kv_cache/fill_cache.cpp (hunk 2 above)
Comment (svuckovicTT) [reply runtime/lib/ttnn/operations/kv_cache/fill_cache.cpp]: Stefan is adding this explicitly to the guidelines, but in the meantime, I'll copy paste it here:

Include style is defined by LLVM coding standards: https://llvm.org/docs/CodingStandards.html#include-style

We prefer these #includes to be listed in this order:

	1. Main Module Header
	2. Local/Private Headers
	3. LLVM project/subproject headers (clang/..., lldb/..., llvm/..., etc)
	4. System #includes

Using `TTIRToTTNN.cpp` as an example, this is what includes would look like for us:

```
#include "ttmlir/Conversion/TTIRToTTNN/TTIRToTTNN.h"  # main header

#include "ttmlir/Dialect/TT/IR/TTOpsTypes.h"  # these are local/private headers
#include "ttmlir/Dialect/TTNN/Utils/Utils.h"

#include "mlir/Dialect/MemRef/IR/MemRef.h"  # llvm project/subproj headers
#include "llvm/Support/LogicalResult.h"

#include <cstdio>  # system includes
#include <algorithm>
```

**Review by mmanzoorTT — APPROVED** [review]

File: include/ttmlir/Dialect/TT/IR/TTOpsTypes.h (hunk 6 above)
Comment (LPanosTT) [reply include/ttmlir/Dialect/TT/IR/TTOpsTypes.h:143]: Oh got it. I'll revert that then.

**Comment by LPanosTT** [comment]
> Hey Lewis this looks great! I presume the `deallocate` of the kv cache is still an issue. How are we working around that?

Well, our desire for a specific graph input to not be deallocated is decoupled from this operations bringup (even though we'll probably never want to deallocate it in the compiled graph). But I still think we'll want to make a solid plan about how we want to define which graph inputs "persist" or not.

File: runtime/lib/ttnn/operations/kv_cache/fill_cache.cpp (hunk 2 above)
Comment (LPanosTT) [reply runtime/lib/ttnn/operations/kv_cache/fill_cache.cpp]: Can we get this in pre-commit?

**Review by jserbedzijaTT — APPROVED** [review]

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 7]
```diff
@@ -1394,6 +1394,112 @@ ::mlir::LogicalResult mlir::tt::ttir::ScatterOp::verify() {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// UpdateCacheOp
+//===----------------------------------------------------------------------===//
+
+::mlir::LogicalResult mlir::tt::ttir::UpdateCacheOp::verify() {
```
[the remark this answers is not in the document]
Comment (LPanosTT) [reply lib/Dialect/TTIR/IR/TTIROps.cpp:1401]: Right now we're only going to support single-batch. I'll assert this value is 0 for now.

**Pull request merged**